File: src/main/java/com/timepath/steam/net/MasterServer.java
package com.timepath.steam.net;

import com.timepath.DataUtils;
import com.timepath.Utils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author TimePath
 * @see <a>https://developer.valvesoftware.com/wiki/Master_Server_Query_Protocol</a>
 */
public class MasterServer extends Server {

    public static final  MasterServer SOURCE = new MasterServer("hl2master.steampowered.com", 27011);
    private static final Logger       LOG    = Logger.getLogger(MasterServer.class.getName());

    public MasterServer(String hostname, int port) {
        super(hostname, port);
    }

    public void query(Region r, com.timepath.steam.net.ServerListener l) throws IOException {
        query(r, "", l);
    }

    public void query(Region r, String filter, com.timepath.steam.net.ServerListener l) throws IOException {
        if(l == null) {
            l = com.timepath.steam.net.ServerListener.NULL;
        }
        String initialAddress = "0.0.0.0:0";
        String lastAddress = initialAddress;
        boolean looping = true;
        while(looping) {
            LOG.log(Level.FINE, "Last address: {0}", lastAddress);
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            baos.write(0x31);
            baos.write(r.code);
            baos.write(( lastAddress + '\0' ).getBytes("UTF-8"));
            baos.write(( filter + '\0' ).getBytes("UTF-8"));
            ByteBuffer send = ByteBuffer.wrap(baos.toByteArray());
            send(send);
            ByteBuffer buf = get();
            int header = buf.getInt();
            if(header != -1) {
                LOG.log(Level.WARNING, "Invalid header {0}", header);
                break;
            }
            byte head = buf.get();
            if(head != 0x66) {
                LOG.log(Level.WARNING, "Unknown header {0}", head);
                String rec = DataUtils.hexDump(buf);
                LOG.log(Level.WARNING, "Received {0}", rec);
                l.inform(rec);
                break;
            }
            byte newline = buf.get();
            if(newline != 0x0A) {
                LOG.log(Level.WARNING, "Malformed byte {0}", newline);
                break;
            }
            int[] octet = new int[4];
            do {
                octet[0] = buf.get() & 0xFF;
                octet[1] = buf.get() & 0xFF;
                octet[2] = buf.get() & 0xFF;
                octet[3] = buf.get() & 0xFF;
                int serverPort = buf.getShort() & 0xFFFF;
                lastAddress = String.format("%d.%d.%d.%d:%d", octet[0], octet[1], octet[2], octet[3], serverPort);
                //noinspection AssignmentUsedAsCondition
                if(looping = !initialAddress.equals(lastAddress)) {
                    l.inform(lastAddress);
                }
            } while(buf.remaining() >= 6);
            if(buf.remaining() > 0) {
                byte[] under = new byte[buf.remaining()];
                if(under.length > 0) {
                    LOG.log(Level.WARNING, "{0} byte underflow: {0}", new Object[] {
                            buf.remaining(), Utils.hex(under)
                    });
                }
            }
        }
    }

    public static enum Region {
        ALL(255),
        US_EAST(0),
        US_WEST(1),
        SOUTH_AMERICA(2),
        EUROPE(3),
        ASIA(4),
        AUSTRALIA(5),
        MIDDLE_EAST(6),
        AFRICA(7);
        private final byte code;

        private Region(int code) {
            this.code = (byte) code;
        }
    }
}

File: src/main/kotlin/com/timepath/steam/net/MasterServer.kt
package com.timepath.steam.net

import com.timepath.DataUtils
import com.timepath.Logger
import com.timepath.Utils

import java.io.IOException
import java.nio.ByteBuffer
import java.util.logging.Level

/**
 * @see <a>https://developer.valvesoftware.com/wiki/Master_Server_Query_Protocol</a>
 */
public class MasterServer(hostname: String, port: Int) : Server(hostname, port) {

    throws(IOException::class)
    public fun query(region: Region, filter: String = "", listener: ServerListener = ServerListener.NULL) {
        val initialAddress = "0.0.0.0:0"
        val filterBytes = filter.toByteArray()
        val write = ByteBuffer.allocate(0
                + /* header */ 1
                + /* region */ 4
                + /* ip - arbitrary */ 30
                + /* filter */ (filterBytes.size() + 1)
        )
        var address = initialAddress
        outer@
        while (true) {
            LOG.log(Level.FINE, { "Last address: ${address}" })
            with(write) {
                reset()
                put(0x31)
                putInt(region.code.toInt())
                put(address.toByteArray())
                put(0)
                put(filterBytes)
                put(0)
                flip()
            }
            send(write)
            val recv = get()
            val header = recv.getInt()
            if (header != -1) {
                LOG.log(Level.WARNING, { "Invalid header ${header}" })
                break
            }
            val head = recv.get()
            if (head.toInt() != 0x66) {
                LOG.log(Level.WARNING, { "Unknown header ${head}" })
                val rec = DataUtils.hexDump(recv)
                LOG.log(Level.WARNING, { "Received ${rec}" })
                listener.inform(rec)
                break
            }
            val newline = recv.get()
            if (newline.toInt() != 0x0A) {
                LOG.log(Level.WARNING, { "Malformed byte ${newline}" })
                break
            }
            do {
                val o0 = recv.get().toInt() and 0xFF
                val o1 = recv.get().toInt() and 0xFF
                val o2 = recv.get().toInt() and 0xFF
                val o3 = recv.get().toInt() and 0xFF
                val port = recv.getShort().toInt() and 0xFFFF
                address = "$o0.$o1.$o2.$o3:$port"
                if (address == initialAddress) {
                    break@outer
                }
                listener.inform(address)
            } while (recv.remaining() >= 6)
            if (recv.capacity() - recv.position() > 0) {
                val under = recv.slice()
                LOG.log(Level.WARNING, { "${recv.remaining()} byte underflow: ${Utils.hex(*under.array())}" })
            }
        }
    }

    public enum class Region private constructor(val code: Byte) {
        ALL(0xFF.toByte()),
        US_EAST(0),
        US_WEST(1),
        SOUTH_AMERICA(2),
        EUROPE(3),
        ASIA(4),
        AUSTRALIA(5),
        MIDDLE_EAST(6),
        AFRICA(7)
    }

    companion object {
        public val SOURCE: MasterServer = MasterServer("hl2master.steampowered.com", 27011)
        private val LOG = Logger()
    }
}
